### PR TITLE
sync Miri

### PR DESCRIPTION
**Pull request by RalfJung**
Created via
```
(cd src/tools && rm -rf miri && git clone https://github.com/rust-lang/miri --depth 1 && rm -rf miri/.git && git add miri)
```

r? @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
@bors r=oli-obk
(I also removed installing xargo from bootstrap)

Do we still need p=1? Miri isn't broken, after all.

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -494,16 +494,6 @@ impl Step for Miri {
...
-        let mut cargo =
-            builder.cargo(compiler, Mode::ToolRustc, SourceType::Submodule, host, "install");
-        cargo.arg("xargo");
-        // Configure `cargo install` path. cargo adds a `bin/`.
-        cargo.env("CARGO_INSTALL_ROOT", &builder.out);
-
-        let mut cargo = Command::from(cargo);
-        if !try_run(builder, &mut cargo) {
-            return;
-        }
```
Comment (oli-obk) [inline src/bootstrap/test.rs:506]: 🎉

**Comment by RalfJung** [comment]
Of course the next josh sync from rustc to Miri will duplicate this commit there...

**Comment by oli-obk** [comment]
> Do we still need p=1? Miri isn't broken, after all.

may be less annoying to resolve conflicts. Not sure what the policy on clippy syncs is

**Comment by RalfJung** [comment]
@bors r-
Investigating  CI failures

**Comment by RalfJung** [comment]
Argh, people changed their licenses...
```
tidy error: invalid license `0BSD` in `enum-iterator 1.2.0 (registry+[https://github.com/rust-lang/crates.io-index)`](https://github.com/rust-lang/crates.io-index)%60)
tidy error: invalid license `0BSD` in `enum-iterator-derive 1.1.0 (registry+[https://github.com/rust-lang/crates.io-index)`](https://github.com/rust-lang/crates.io-index)%60)
tidy error: invalid license `(MIT OR Apache-2.0) AND Unicode-DFS-2016` in `unicode-ident 1.0.4 (registry+[https://github.com/rust-lang/crates.io-index)`](https://github.com/rust-lang/crates.io-index)%60)
tidy error: Dependencies for main workspace not explicitly permitted:
* unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)
```

**Comment by RalfJung** [comment]
Maybe we should just get rid of vergen, that would help with some of these. clippy seems to use a [local fork of](https://github.com/rust-lang/rust-clippy/tree/master/rustc_tools_util) https://crates.io/crates/rustc_tools_util. @oli-obk any idea why clippy uses a fork rather than fetching it from crates.io?

**Comment by matthiaskrgr** [comment]
@RalfJung the original source of the crate lives in the clippy repo, it shouldn't be a fork :D

**Comment by RalfJung** [comment]
Yeah I realized it's not a fork, but if Miri now uses the crates.io version then we have code duplication in the rustc workspace. Also the crates.io version is outdated but I don't know if the diff is relevant.

Cc https://github.com/rust-lang/rust-clippy/issues/9553

**Comment by RalfJung** [comment]
The uncide-ident error is strange -- this is a cargo-miri dependency, but we get an error for it supposedly being a rustc dependency?

**Comment by RalfJung** [comment]
However if we avoid vergen I think we can also avoid updating syn and then this license becomes someone else's problem...^^

**Comment by RalfJung** [comment]
> Of course the next josh sync from rustc to Miri will duplicate this commit there...

... and this does seem to lead to conflicts down the road, when I try to sync rustc back into Miri after this PR.

I am not quite sure how to proceed here. This might be fine for a one-off change, but longer term we probably want to get josh to work for both directions, which I guess would work better if we had done the initial subtree merge with josh instead of git-subtree... but now we need to figure out how to live with the repo that we have.

**Comment by oli-obk** [comment]
Well... the proper josh way would be to erase all miri history here, us not doing that is where the failure to sync miri->rustc is coming from.

Alternatively... we can re-sync the src/tools/miri folder from scratch... creating *another* duplication of all miri commits in the rustc repo 🙈

**Comment by RalfJung** [comment]
> Well... the proper josh way would be to erase all miri history here, us not doing that is where the failure to sync miri->rustc is coming from.

How can we erase history in git?

> Alternatively... we can re-sync the src/tools/miri folder from scratch... creating another duplication of all miri commits in the rustc repo see_no_evil

Is it *literally* the original commits? Because then it wouldn't duplicate, it would reference the same commits that git-subtree already added to the repo.

**Comment by oli-obk** [comment]
> Is it _literally_ the original commits? Because then it wouldn't duplicate, it would reference the same commits that git-subtree already added to the repo.

hmm... I don't believe it works the same way, no. josh rewrites the commits afaict.

**Comment by RalfJung** [comment]
Okay so I think the reason syncing usually just 'magically' works in josh is that when we sync Miri changes into rustc, and then export them to Miri again again, they end up with the original SHA. So then syncing them into Miri again should work neatly.

Using josh only for one direction will not work well I think.

**Comment by RalfJung** [comment]
I got josh to create a sync, pushed to alternative PR https://github.com/rust-lang/rust/pull/102573.

**Pull request closed without merging**